std: optimize `dlsym!` macro and add a test for it

joboet opened the pull request:
The `dlsym!` macro always ensures that the name string is nul-terminated, so there is no need to perform the check at runtime. Also, acquire loads are generally faster than a load and a barrier, so use them. This is only false in the case where the symbol is missing, but that shouldn't matter too much.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17335095555/job/49219709447) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49219709447) [(plain)](https://github.com/rust-lang/rust/commit/3aee6a7906d2f69536be7dddd0356760b7e12830/checks/49219709447/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error[E0308]: mismatched types
   --> library/std/src/sys/pal/unix/weak.rs:100:13
    |
100 |             name,
    |             ^^^^ expected `*const u8`, found `&CStr`
    |
    = note: expected raw pointer `*const u8`
                 found reference `&CStr`

For more information about this error, try `rustc --explain E0308`.
```

</details>

joboet committed:
std: improve the `dlsym!` macro and add a test for it

* Ensure nul-termination of the symbol name at compile-time
* Use an acquire load instead of a relaxed load and acquire fence
* Properly use `unsafe` and add safety comments
* Add tests

rust-timer commented:
Queued e2bfd7fb1f49deb317bdee4685858a4b7e85d848 with parent fe55364329579d361b1ab565728bc033a7dba07e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=fe55364329579d361b1ab565728bc033a7dba07e&end=e2bfd7fb1f49deb317bdee4685858a4b7e85d848).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e2bfd7fb1f49deb317bdee4685858a4b7e85d848): [comparison URL](https://perf.rust-lang.org/compare.html?start=fe55364329579d361b1ab565728bc033a7dba07e&end=e2bfd7fb1f49deb317bdee4685858a4b7e85d848&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=fe55364329579d361b1ab565728bc033a7dba07e&end=e2bfd7fb1f49deb317bdee4685858a4b7e85d848&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.2%, -0.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.9%, -2.5%] | 6     |
| All ❌✅ (primary)                 | -0.2% | [-0.2%, -0.2%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=fe55364329579d361b1ab565728bc033a7dba07e&end=e2bfd7fb1f49deb317bdee4685858a4b7e85d848&stat=max-rss)
<details>
<summary>Results (primary -1.0%, secondary -1.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0%  | [2.7%, 3.3%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -9.0% | [-9.0%, -9.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.9% | [-1.9%, -1.9%] | 1     |
| All ❌✅ (primary)                 | -1.0% | [-9.0%, 3.3%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=fe55364329579d361b1ab565728bc033a7dba07e&end=e2bfd7fb1f49deb317bdee4685858a4b7e85d848&stat=cycles:u)
<details>
<summary>Results (primary 1.9%, secondary 2.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.9% | [1.9%, 1.9%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.9% | [1.9%, 1.9%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=fe55364329579d361b1ab565728bc033a7dba07e&end=e2bfd7fb1f49deb317bdee4685858a4b7e85d848&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 467.034s -> 466.989s (-0.01%)
**Artifact size**: 388.52 MiB -> 388.52 MiB (-0.00%)

joboet commented on library/std/src/sys/pal/unix/weak.rs:
```diff
@@ -78,62 +81,65 @@ pub(crate) macro dlsym {
+            if size_of::<F>() != size_of::<*mut libc::c_void>() {
+                panic!("not a function pointer")
+            }
+        }
```
> Optional, but this could probably be enforced a bit stronger with the bound `F: crate::marker::FnPtr`. Even if that bound is added, we should likely avoid using its `.addr()` method for now until it gets the planned updates (maybe worth a FIXME).

Makes sense, done!

> Also, it's preexisting but any idea why `transmute_copy` is used rather than `transmute`? That would enforce the size constraint.

`transmute` complains about `F` being generic...

tgross35 reviewed (approved):
Looks great, r=me with a squash

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146019/fe55364329579d361b1ab565728bc033a7dba07e..0332a66a25fa6a90cbad7c5f0788d5697b811cef)*

joboet commented:
@bors r=@tgross35

bors commented:
:pushpin: Commit d0a27614d1dfa14094a50aa3a930b633c9532056 has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 d0a27614d1dfa14094a50aa3a930b633c9532056 -->
<!-- homu: {"type":"Approved","sha":"d0a27614d1dfa14094a50aa3a930b633c9532056","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d0a27614d1dfa14094a50aa3a930b633c9532056 with merge 0b56db014aab3d06cc0aaf1699ac48b6562f2b1c...
<!-- homu: {"type":"BuildStarted","head_sha":"d0a27614d1dfa14094a50aa3a930b633c9532056","merge_sha":"0b56db014aab3d06cc0aaf1699ac48b6562f2b1c"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17419103266/job/49453846359) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49453846359) [(plain)](https://github.com/rust-lang/rust/commit/0b56db014aab3d06cc0aaf1699ac48b6562f2b1c/checks/49453846359/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- sys::pal::unix::weak::tests::dlsym_existing stdout ----

thread 'sys::pal::unix::weak::tests::dlsym_existing' (515653) panicked at library/std/src/sys/pal/unix/weak/tests.rs:16:25:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---- sys::pal::unix::weak::tests::dlsym_existing stdout end ----

failures:
    sys::pal::unix::weak::tests::dlsym_existing

test result: FAILED. 542 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 3.54s
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/17419103266/job/49457419107)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/17419103266/job/49457419107","builder_name":"checks-actions"} -->

joboet committed:
std: only test `dlsym!` on platforms where it is actually used

`dlsym` doesn't work for finding libc symbols on platforms like linux-musl, so the test will fail.

joboet commented:
@bors r=@tgross35
